src/users/users.service.ts

```diff
@@ -0,0 +1,91 @@
+import {
+  ForbiddenException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
+import { PrismaService } from '../database/prisma.service';
+import { CreateUserDto } from '../auth/input/auth.dto';
+import * as bcrypt from 'bcrypt';
+import { PrismaClientKnownRequestError } from '@prisma/client/runtime/library';
+import { DatabaseRecord } from '../common/models/databaseRecord.model';
+import { User } from './models/user.type';
+import { MailService } from '../mail/mail.service';
+
+@Injectable()
+export class UsersService {
+  constructor(
+    private prisma: PrismaService,
+    private mailService: MailService,
+  ) {}
+
+  public async createUser(
+    userDto: CreateUserDto,
+  ): Promise<DatabaseRecord<User>> {
+    const salt: string = await bcrypt.genSalt();
+    const hash: string = await bcrypt.hash(userDto.password, salt);
+
+    const user: DatabaseRecord<User> = await this.prisma.user
+      .create({
+        data: {
+          email: userDto.email,
+          name: userDto.name,
+          hash,
+        },
+      })
+      .then((user) => {
+        this.mailService.sendUserConfirmation(user);
+        return user;
+      })
+      .catch((error) => {
+        if (error instanceof PrismaClientKnownRequestError) {
+          throw new ForbiddenException('Credentials incorrect');
+        }
+        throw error;
+      });
```

Comment:
hmm, I kind of don't like mixing `awaits` with `Promise.then`. I would change this a bit:

```suggestion
  try {
    const user: DatabaseRecord<User> = await this.prisma.user
      .create({
        data: {
          email: userDto.email,
          name: userDto.name,
          hash,
        },
      });
      
      this.mailService.sendUserConfirmation(user);
  } catch (error) {
      if (error instanceof PrismaClientKnownRequestError) {
        throw new ForbiddenException('Credentials incorrect');
      }
        
      throw error;
  }
```